### PR TITLE
feat(test): Add support to run UI tests in CI pipeline

### PR DESCRIPTION
**Pull request by jagan-jaya**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Adding UI tests to CI pipeline

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

To improve our system reliability, we needs UI tests to check the redirection scenarios and wallet scenarios.

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code

### CODE REVIEW

File: scripts/run_ui_tests.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+echo 'Unattended-Upgrade::Allowed-Origins:: "LP-PPA-mozillateam:${distro_codename}";' | sudo tee /etc/apt/apt.conf.d/51unattended-upgrades-firefox
+sudo apt install -y firefox
+firefox
+
+#start server and run ui tests
+cargo run &
+
+#Wait for the server to start in port 8080
+while netstat -lnt | awk '$4 ~ /:8080$/ {exit 1}'; do sleep 10; done
+cargo test --package router --test connectors -- $1"_ui::" --test-threads=1 >> tests/test_results.log 2>&1
```
Comment (SanchithHegde) [inline scripts/run_ui_tests.sh:48]: Can't we directly print logs to stdout instead? That also gives us realtime info about the progress of tests.

And as far as I understand, it should also handle failing the workflow if the tests fail.

File: scripts/run_ui_tests.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+Package: * 
+Pin: release o=LP-PPA-mozillateam 
+Pin-Priority: 1001 
+' | sudo tee /etc/apt/preferences.d/mozilla-firefox
+echo 'Unattended-Upgrade::Allowed-Origins:: "LP-PPA-mozillateam:${distro_codename}";' | sudo tee /etc/apt/apt.conf.d/51unattended-upgrades-firefox
+sudo apt install -y firefox
+firefox
+
+#start server and run ui tests
+cargo run &
```
Comment (SanchithHegde) [inline scripts/run_ui_tests.sh:44]: Can we avoid running this in the background?

How are we handling failures in this command, either compilation itself failing or application not starting due to incorrect configuration?

File: scripts/run_ui_tests.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
+sudo apt update
+apt install net-tools
+apt-get install wget
```
Comment (SanchithHegde) [inline scripts/run_ui_tests.sh:3]: We're performing Ubuntu-specific operations, we should either move this to the `.github` directory, or make this script suitable to work with Unixes.

File: scripts/run_ui_tests.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# "CREATE USER $DB_USER WITH PASSWORD '$DB_PASS' SUPERUSER CREATEDB CREATEROLE INHERIT LOGIN;"
+# sudo -u postgres psql -e -c \
+# "CREATE DATABASE $DB_NAME;"
+# diesel migration --database-url postgres://db_user:db_pass@localhost:5432/hyperswitch_db run
+
+#install geckodriver
+wget -c https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux-aarch64.tar.gz && tar -xvzf geckodriver*
+chmod +x geckodriver
+mv geckodriver /usr/local/bin/
+geckodriver > tests/geckodriver.log 2>&1 &
```
Comment (SanchithHegde) [inline scripts/run_ui_tests.sh:30]: Are there any available GitHub Actions for installing geckodriver and firefox headless?

File: scripts/run_ui_tests.sh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# service postgresql start &
+# cargo install diesel_cli --no-default-features --features "postgres"
+# export DB_USER="db_user"
+# export DB_PASS="db_pass"
+# export DB_NAME="hyperswitch_db"
+# sudo -u postgres psql -e -c \
+# "CREATE USER $DB_USER WITH PASSWORD '$DB_PASS' SUPERUSER CREATEDB CREATEROLE INHERIT LOGIN;"
+# sudo -u postgres psql -e -c \
+# "CREATE DATABASE $DB_NAME;"
+# diesel migration --database-url postgres://db_user:db_pass@localhost:5432/hyperswitch_db run
```
Comment (SanchithHegde) [inline scripts/run_ui_tests.sh:24]: Remove this.

File: crates/router/tests/connectors/selenium.rs [hunk 6]
```diff
@@ -481,8 +493,13 @@ pub fn make_capabilities(s: &str) -> Capabilities {
...
-            let profile_path = &format!("-profile={}", get_firefox_profile_path().unwrap());
-            caps.add_firefox_arg(profile_path).unwrap();
+            let ignore_profile = env::var("IGNORE_BROWSER_PROFILE").ok();
+            if ignore_profile.is_none() {
+                let profile_path = &format!("-profile={}", get_firefox_profile_path().unwrap());
+                caps.add_firefox_arg(profile_path).unwrap();
+            } else {
+                caps.add_firefox_arg("--headless").ok();
+            }
             // let mut prefs = FirefoxPreferences::new();
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/selenium.rs:503]: Do we need the commented code under this and the `chrome` arm?

File: .github/workflows/connector-ui-sanity-tests.yml [hunk 7]
```diff
@@ -0,0 +1,105 @@
+name: Connector UI Sanity Tests
+
+on:
+  push:
+    branches:
+      - main
+  
+  pull_request:
```
[the remark this answers is not in the document]
Comment (jarnura) [reply .github/workflows/connector-ui-sanity-tests.yml:7]: If all cases where working we can remove this check. Merge queue will run these cases.

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

File: .github/scripts/run_ui_tests.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+Package: * 
+Pin: release o=LP-PPA-mozillateam 
+Pin-Priority: 1001 
+' | sudo tee /etc/apt/preferences.d/mozilla-firefox
+echo 'Unattended-Upgrade::Allowed-Origins:: "LP-PPA-mozillateam:${distro_codename}";' | sudo tee /etc/apt/apt.conf.d/51unattended-upgrades-firefox
+sudo apt install -y firefox
+firefox
+
+#start server and run ui tests
+cargo run &
```
Comment (ashokkjag) [inline .github/scripts/run_ui_tests.sh:27]: these lines will wait indefinitely if there is a build failure. rather use cargo build run the binary in background. Even with this. Just make sure below wait loop exits after reasonable time.

File: .github/scripts/run_ui_tests.sh [hunk 9]
```diff
@@ -0,0 +1,30 @@
+sudo apt update
+apt install net-tools
+apt-get install wget
+mkdir tests
+
+#download connector ui tests
+mv .github/testcases/ui_tests.json $HOME/target/test/connector_tests.json
+
+#install geckodriver
+wget -c https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux-aarch64.tar.gz && tar -xvzf geckodriver*
```
Comment (ashokkjag) [inline .github/scripts/run_ui_tests.sh:10]: The github runner images already have firefox/gecko drivers. Can you check if they are sufficient?

https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#browsers-and-drivers

```
Selenium server 4.10.0
Mozilla Firefox 114.0.2
Geckodriver 0.33.0
```

**Review by ashokkjag — CHANGES_REQUESTED** [review]

**Review by ashokkjag — DISMISSED** [review]

File: .github/scripts/run_ui_tests.sh [hunk 10]
```diff
@@ -0,0 +1,22 @@
+sudo apt update
```
Comment (SanchithHegde) [inline .github/scripts/run_ui_tests.sh:1]: Add a shebang: `#! /usr/bin/env bash`

File: crates/router/tests/connectors/selenium.rs [hunk 11]
```diff
@@ -47,13 +46,28 @@ pub enum Assert<'a> {
...
-pub static CHEKOUT_DOMAIN: &str = "hs-payments-test.netlify.app";
 #[async_trait]
 pub trait SeleniumTest {
+    fn get_saved_testcases(&self) -> serde_json::Value {
+        let env_value = env::var("CONNECTOR_TESTS_FILE_PATH").ok();
+        if env_value.is_none() {
+            return serde_json::json!("");
+        }
+        let path =
+            env::var("CONNECTOR_TESTS_FILE_PATH").expect("connector tests file path not set");
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/selenium.rs:62]: What exactly is the intended behavior?

Comment (jagan-jaya) [reply crates/router/tests/connectors/selenium.rs:62]: Connector testcases are maintained outside the repo, if this path is not provided manually the UI tool will try to download the testcases each time we run the tests, we are avoiding the extra call here

**Review by SanchithHegde — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**